Upload election definition

cikzh opened the pull request:
<!--
- What's the scope of the changes?
- What did you test? Which edge cases did you explicitly take into account?
- Are there things you want reviewers to definitely take a look at?
- Are there specific people you want feedback from?
- Do you have tips on how to test? (For example: data setup, triggering specific errors, etc.)
-->

cikzh commented on backend/src/eml/hash.rs:
```diff
@@ -0,0 +1,93 @@
+}
+
+#[derive(Debug, Deserialize, Serialize, Clone, ToSchema)]
+pub struct RetractedEmlHash {
```
Renamed in [b920a5f](https://github.com/kiesraad/abacus/pull/1406/commits/b920a5f795a45186f53921fc808a24399857c6da)

jorisleker reviewed (changes requested):
Tested, good start!

**Not as designed/expected**:
- spacing between first and second line of hash is bigger then design
- the alert that is shown when uploading an invalid file, should span the entire right column instead of just the width of the text paragraphs.
- interaction: expected behavior is that **only** the part of the hash that is linked to the focussed input, is highlighted with interactive default color
- upload a file, click 'volgende' and then hit the back button. Observed: you navigate back to /elections. Expected: upload form is shown again. This will become a bigger issue when the other steps in the wizard are implemented. Back button is expected to navigate back 1 step, instead of navigating away from the wizard.

**Design updated based on this review**
- I've updated the design so the '1' and '2' blocks are exactly the same width as 4 monospaced chars. Bit nicer now the hash spans two rows. Can you update css as well?
- I've added the name of the uploaded file to the error alert when uploading an invalid file, so it's easier to spot the mistake you made.

**Question / to be discussed:**
- If I upload the same file multiple times, a different part of the hash is masked. Is this intended behavior? This way, I can circumvent the check by trying to upload the file multiple times. Shouldn't we always mask the same two blocks for a given file (ie by using the hash as seed for random()?)

Lionqueen94 commented on backend/src/election/api.rs:
```diff
@@ -0,0 +1,155 @@
+use axum::http::StatusCode;
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+use utoipa_axum::router::OpenApiRouter;
+use utoipa_axum::routes;
```
```suggestion
use utoipa_axum::{router::OpenApiRouter, routes};
```

cikzh replied:
Not sure why rustfmt doesn't format this... Fixed in [882d161](https://github.com/kiesraad/abacus/pull/1406/commits/882d161a16951a3a57a736487ec5ee9d6e194068)

Lionqueen94 commented on backend/src/election/api.rs:
```diff
@@ -0,0 +1,155 @@
+#[cfg(feature = "dev-database")]
+use crate::audit_log::{AuditEvent, AuditService};
+use crate::authentication::{Admin, User};
+use crate::election::ElectionRequest;
+use crate::election::repository::Elections;
```
```suggestion
use crate::election::{ElectionRequest, repository::Elections};
```

cikzh replied:
Not sure why rustfmt doesn't format this... Fixed in [882d161](https://github.com/kiesraad/abacus/pull/1406/commits/882d161a16951a3a57a736487ec5ee9d6e194068)

Lionqueen94 commented on backend/src/election/api.rs:
```diff
@@ -0,0 +1,155 @@
+use crate::election::ElectionRequest;
+use crate::election::repository::Elections;
+use crate::eml::{EML110, EMLDocument, RedactedEmlHash};
+use crate::polling_station::PollingStation;
+use crate::polling_station::repository::PollingStations;
```
```suggestion
use crate::polling_station::{PollingStation, repository::PollingStations};
```

cikzh replied:
Not sure why rustfmt doesn't format this... Fixed in [882d161](https://github.com/kiesraad/abacus/pull/1406/commits/882d161a16951a3a57a736487ec5ee9d6e194068)

Lionqueen94 commented on backend/src/eml/common.rs:
```diff
@@ -146,6 +150,22 @@ pub struct ElectionDomain {
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum EMLImportError {
+    MissingRegion,
+    OnlyMunicipalSupported,
+    MissingNumberOfSeats,
+    NumberOfSeatsNotInRange,
+    InvalidDateFormat,
+    MissingNominationDate,
+    MissingSubcategory,
+    MismatchNumberOfSeats,
+    MissingPreferenceThreshold,
+    MismatchPreferenceThreshold,
+    Needs101a,
+    TooManyPoliticalGroups,
+}
```
I think it would be nice to make this list in alphabetical order

cikzh replied:
Fixed in  [654bea8](https://github.com/kiesraad/abacus/pull/1406/commits/654bea8f24cefc6110505c0b70c8801631378387)

Lionqueen94 commented on backend/src/error.rs:
```diff
@@ -31,6 +32,8 @@ pub enum ErrorReference {
     DrawingOfLotsRequired,
     EntryNotFound,
     EntryNotUnique,
+    EntryNumberNotSupported,
```
This was removed in a merged PR since it was not used, please remove it again and regenerate openapi.json :)

cikzh replied:
Fixed in [c443ac5](https://github.com/kiesraad/abacus/pull/1406/commits/c443ac5da7ea72390f4d4185566cbb0d9bc571d1)

Lionqueen94 commented on frontend/src/components/ui/Alert/Alert.tsx:
```diff
@@ -12,18 +12,28 @@ import cls from "./Alert.module.css";
   return (
     <div
-      className={cn(inline ? cls.inlineAlert : cls.alert, cls[type], margin, variant)}
+      className={cn(inline ? cls.inlineAlert : cls.alert, cls[type], margin, variant, small ? "small" : undefined)}
```
```suggestion
      className={cn(inline ? cls.inlineAlert : cls.alert, cls[type], margin, variant, { small: small })}
```

Lionqueen94 commented on frontend/src/lib/i18n/locales/nl/election.json:
```diff
@@ -1,6 +1,15 @@
 {
   "create": "Verkiezing toevoegen",
   "import_eml": "Importeer verkiezingsdefinitie",
+  "check_eml": {
+    "title": "Controleer verkiezingsdefinitie",
+    "description": "Het bestand <file>filename</file> is geïmporteerd en bevat de instellingen voor de volgende verkiezing:",
+    "check_hash": {
+      "description": "Vul ter controle de hierboven afgeschermde delen van de digitale vingerafdruk in. De volledige code heb je van de Kiesraad ontvangen.",
+      "label": "Controle deel {stub}",
+      "hint": "Vul 4 karakters in"
+    }
+  },
```
```suggestion
  "check_eml": {
    "title": "Controleer verkiezingsdefinitie",
    "description": "Het bestand <file>filename</file> is geïmporteerd en bevat de instellingen voor de volgende verkiezing:",
    "check_hash": {
      "description": "Vul ter controle de hierboven afgeschermde delen van de digitale vingerafdruk in. De volledige code heb je van de Kiesraad ontvangen.",
      "label": "Controle deel {stub}",
      "hint": "Vul 4 karakters in"
    }
  },
  "import_eml": "Importeer verkiezingsdefinitie",
  "invalid_election_definition": {
    "title": "Ongeldige verkiezingsdefinitie",
    "description": "Het gekozen bestand bevat geen geldige verkiezingsdefinitie. Kies een bestand met een geldige definitie."
  },
```

Lionqueen94 commented on frontend/src/lib/i18n/locales/nl/election.json:
```diff
@@ -16,5 +25,9 @@
     "singular": "Verkiezing"
   },
-  "use_instructions_to_import_eml": "Je hebt van de Kiesraad instructies gekregen waarmee je de verkiezingsdefinitie kunt downloaden. Zet het bestand op deze computer en importeer het."
+  "use_instructions_to_import_eml": "Je hebt van de Kiesraad instructies gekregen waarmee je de verkiezingsdefinitie kunt downloaden. Zet het bestand op deze computer en importeer het.",
+  "invalid_election_definition": {
+    "title": "Ongeldige verkiezingsdefinitie",
+    "description": "Het gekozen bestand bevat geen geldige verkiezingsdefinitie. Kies een bestand met een geldige definitie."
+  }
```
```suggestion
  "use_instructions_to_import_eml": "Je hebt van de Kiesraad instructies gekregen waarmee je de verkiezingsdefinitie kunt downloaden. Zet het bestand op deze computer en importeer het."
```

Lionqueen94 commented on frontend/src/lib/i18n/locales/nl/error.json:
```diff
@@ -21,9 +21,12 @@
     "DrawingOfLotsRequired": "Loting is noodzakelijk, maar nog niet beschikbaar in deze versie van Abacus",
     "EntryNotFound": "Niet gevonden",
     "EntryNotUnique": "Deze invoer bestaat al",
+    "EntryNumberNotSupported": "Ongeldige invoer voor dit stembureau",
```
```suggestion
```

cikzh replied:
Fixed in [c443ac5](https://github.com/kiesraad/abacus/pull/1406/commits/c443ac5da7ea72390f4d4185566cbb0d9bc571d1)

cikzh commented:
> **Not as designed/expected**:
> 
>     * spacing between first and second line of hash is bigger then design
Fixed in [9355969](https://github.com/kiesraad/abacus/pull/1406/commits/9355969a752878588ce3bce7e075fb0c36edfcbd)
>     * the alert that is shown when uploading an invalid file, should span the entire right column instead of just the width of the text paragraphs.
Fixed in  [07a4528](https://github.com/kiesraad/abacus/pull/1406/commits/07a45280d8dd5bd6bbecb1a5296e98ecdfc51834)
>     * interaction: expected behavior is that **only** the part of the hash that is linked to the focussed input, is highlighted with interactive default color
Hmm... I intended it that way, but introduced a bug somewhere. I'll fix it.
>     * upload a file, click 'volgende' and then hit the back button. Observed: you navigate back to /elections. Expected: upload form is shown again. This will become a bigger issue when the other steps in the wizard are implemented. Back button is expected to navigate back 1 step, instead of navigating away from the wizard.
I'll keep this out of scope of this PR. In the next issue(s) of this epic I'm going to implement the navigation logic for the whole page, so I'll implement it there
> 
> **Design updated based on this review**
> 
>     * I've updated the design so the '1' and '2' blocks are exactly the same width as 4 monospaced chars. Bit nicer now the hash spans two rows. Can you update css as well?
Fixed in [9355969](https://github.com/kiesraad/abacus/pull/1406/commits/9355969a752878588ce3bce7e075fb0c36edfcbd)
>     * I've added the name of the uploaded file to the error alert when uploading an invalid file, so it's easier to spot the mistake you made.
Fixed in [07a4528](https://github.com/kiesraad/abacus/pull/1406/commits/07a45280d8dd5bd6bbecb1a5296e98ecdfc51834)
> 
> **Question / to be discussed:**
> 
>     * If I upload the same file multiple times, a different part of the hash is masked. Is this intended behavior? This way, I can circumvent the check by trying to upload the file multiple times. Shouldn't we always mask the same two blocks for a given file (ie by using the hash as seed for random()?)
AFAIK the main reason for this check is so the user pays extra attention to the SHA hash, so they do a double check they upload the correct file. If there also security concerns for this check, then it's probably a good idea to put in a little more effort in the backend code so it always returns the same stubs. Let me know what you think and I'm also curious of @praseodym's thoughts on this matter.

Lionqueen94 commented on frontend/src/features/election_management/components/CheckElectionDefinition.tsx:
```diff
@@ -39,10 +41,12 @@ export function CheckElectionDefinition({ file, election, hash }: CheckElectionD
-          <strong>Digitale vingerafdruk</strong> (hashcode):
+        <div>
+          <span>
+            <strong>Digitale vingerafdruk</strong>(hashcode):
```
Please add this as an i18n variable

cikzh replied:
Ah, damn, keep forgetting these. Thanks!

cikzh replied:
Fixed in [f5d1d1b](https://github.com/kiesraad/abacus/pull/1406/commits/f5d1d1ba34f7e0b83bf49378e513b3cf24efbf65)

cikzh commented on frontend/src/features/election_management/components/RedactedHash.module.css:
```diff
@@ -0,0 +1,30 @@
+}
+
+.chunkStub {
+  width: 39px;
```
The alignment has to be pixel perfect, or the two lines get unaligned. I've tried to do it with the `ex` unit, because I'm trying to align it with a monospace font, but that didn't really work.

jorisleker replied:
Alternative to get this work a bit more reliant: 
give all stubs a min-width, and center align the chars inside. That way they can all be `2.5rem`

cikzh replied:
That seems to work, thanks. Fixed in [c26489b](https://github.com/kiesraad/abacus/pull/1406/commits/c26489b22e56b883afabf800b77455c436648eff). I've left out the `text-align: center`, though. All it seemed to do was shift the whole thing right by 1px, making unaligned with the other texts on the page.

cikzh commented on frontend/src/features/election_management/components/UploadElectionDefinition.tsx:
```diff
@@ -0,0 +1,58 @@
+  };
+
+  return (
+    <section className={"md"}>
```
Fixed in [f5d1d1b](https://github.com/kiesraad/abacus/pull/1406/commits/f5d1d1ba34f7e0b83bf49378e513b3cf24efbf65)

cikzh commented on frontend/src/features/election_management/components/CheckElectionDefinition.module.css:
```diff
@@ -0,0 +1,2 @@
+.container {
+}
```
Fixed in [108b3ae](https://github.com/kiesraad/abacus/pull/1406/commits/108b3aee12ecf6a9659f6b4ccb294d6e8853f344)

cikzh commented on frontend/src/features/election_management/components/CheckElectionDefinition.tsx:
```diff
@@ -0,0 +1,86 @@
+import { cn } from "@/utils/classnames";
+import { formatDateFull } from "@/utils/format";
+
+import cls from "./CheckElectionDefinition.module.css";
```
Fixed in [108b3ae](https://github.com/kiesraad/abacus/pull/1406/commits/108b3aee12ecf6a9659f6b4ccb294d6e8853f344)

cikzh commented on frontend/src/features/election_management/components/CheckElectionDefinition.tsx:
```diff
@@ -0,0 +1,86 @@
+  );
+
+  return (
+    <section className={cn("md", cls.container)}>
```
Fixed in [108b3ae](https://github.com/kiesraad/abacus/pull/1406/commits/108b3aee12ecf6a9659f6b4ccb294d6e8853f344)

jorisleker reviewed (changes requested):
One tiny issue remaining.

In the invalid file step, the file upload control should show 'Geen bestand gekozen' instead of the name of the uploaded file.
![Scherm­afbeelding 2025-04-25 om 13 19 43](https://github.com/user-attachments/assets/95f930b6-56ba-41ea-8272-53e435f7dd59)

cikzh commented on backend/src/election/api.rs:
```diff
@@ -0,0 +1,152 @@
+use crate::election::{ElectionRequest, repository::Elections};
+use crate::eml::{EML110, EMLDocument, RedactedEmlHash};
+use crate::polling_station::{PollingStation, repository::PollingStations};
+use crate::{AppState, ErrorResponse};
```
Done in [bb5bc3e](https://github.com/kiesraad/abacus/pull/1406/commits/bb5bc3e11ad4eb2b257a0c1f1f117e08b1b6d7c9)

cikzh commented on frontend/src/utils/format.ts:
```diff
@@ -52,6 +52,15 @@ function isYesterday(date: Date): boolean {
   );
 }
 
+export function formatDateFull(date: Date) {
```
Added in [54134ba](https://github.com/kiesraad/abacus/pull/1406/commits/54134ba67c446b97105f988ef4d1eaa868c8f22a)

cikzh commented on frontend/src/features/election_management/components/ElectionCreatePage.test.tsx:
```diff
@@ -0,0 +1,35 @@
+    expect(await screen.findByRole("heading", { level: 1, name: "Verkiezing toevoegen" }));
+    expect(await screen.findByRole("heading", { level: 2, name: "Importeer verkiezingsdefinitie" }));
+    expect(await screen.findByLabelText("Bestand kiezen"));
+    expect(await screen.findByLabelText("Geen bestand gekozen"));
```
Fixed in [01e6d6d](https://github.com/kiesraad/abacus/pull/1406/commits/01e6d6df44a7b318aa99364c7ee8cbc3642c39b4)

cikzh commented:
> In the invalid file step, the file upload control should show 'Geen bestand gekozen' instead of the name of the uploaded file.
 @jorisleker Fixed in [9a5148b](https://github.com/kiesraad/abacus/pull/1406/commits/9a5148b71cab5f6c4a74c969e2d8d602511bd2a6)
@cikzh

Lionqueen94 commented on frontend/src/lib/i18n/locales/nl/generic.json:
```diff
@@ -17,6 +17,7 @@
   "date_locale": "nl-NL",
   "delete": "Verwijderen",
   "description": "Omschrijving",
+  "digital_signature": "Digital signature",
```
```suggestion
  "digital_signature": "Digitale vingerafdruk",
```

jschuurk-kr commented on frontend/src/features/election_management/components/CheckElectionDefinition.tsx:
```diff
@@ -0,0 +1,84 @@
+        </p>
+        <div>
+          <span>
+            <strong>{t("digital_signature")}</strong>({t("hashcode")}):
```
Missing space, page says: "Digitale vingerafdruk(hashcode):"

```suggestion
            <strong>{t("digital_signature")}</strong> ({t("hashcode")}):
```

jorisleker reviewed (dismissed):
UI LGTM now

jschuurk-kr commented on frontend/src/features/election_management/components/UploadElectionDefinition.tsx:
```diff
@@ -0,0 +1,59 @@
+            <p>
+              {tx("election.invalid_election_definition.description", {
+                file: () => {
+                  return <strong>{error}</strong>;
```
assuming I'm looking at the latest commit via https://abacus-test.nl/ and/or it has not been changed in later commits

jorisleker replied:
Yes, this as designed

Lionqueen94 commented on backend/src/eml/eml_110.rs:
```diff
@@ -272,4 +411,119 @@ mod tests {
+        assert!(matches!(res, EMLImportError::MismatchNumberOfSeats));
+    }
+
+    #[test]
+    fn test_election_validate_invalid_election_subcategory() {
+        let data = include_str!("./tests/eml110a_invalid_election_subcategory.eml.xml");
+        let doc = EML110::from_str(data).unwrap();
+        let res = doc.as_crate_election().unwrap_err();
+        assert!(matches!(res, EMLImportError::MismatchNumberOfSeats));
+    }
```
This test name does not match the assert

Lionqueen94 replied:
You probably want to check for `EMLImportError::OnlyMunicipalSupported` here, since this is still missing in the testcoverage

rnijveld replied:
I've added a comment as to why the assert doesn't match, if you change either the number of seats or the election subcategory then we have no easy way of knowing which one was changed, so we just go for the easy solution and pick one. I've also added the missing test for a non-municipal election definition in [85676ec](https://github.com/kiesraad/abacus/pull/1406/commits/85676ecbc56181f6e7f3051665203c6848fb7098)

Lionqueen94 commented on backend/src/eml/eml_110.rs:
```diff
@@ -272,4 +411,119 @@ mod tests {
+    fn test_election_validate_invalid_election_missing_region() {
+        let data = include_str!("./tests/eml110a_invalid_election_missing_region.eml.xml");
+        let doc = EML110::from_str(data).unwrap_err();
+        assert!(matches!(doc, DeError::Custom(_)));
```
This should check for `EMLImportError::MissingRegion` right?

rnijveld replied:
Nope! This is already captured by the parser, I've added an 'artificially' constructed EML struct just to make sure we cover the branch anyway in [85676ec](https://github.com/kiesraad/abacus/pull/1406/commits/85676ecbc56181f6e7f3051665203c6848fb7098), but it is not something we should see in regular usage the way the parser is set up currently. I did want to check against the empty array though as it would be something we could not handle further on.

praseodym commented on frontend/src/testing/setup.ts:
```diff
@@ -1,5 +1,6 @@
 import "@testing-library/jest-dom/vitest";
 import { cleanup, configure } from "@testing-library/react";
+import "blob-polyfill";
```
Did you try the approach suggested in https://github.com/jsdom/jsdom/issues/2555#issuecomment-1864762292?

rnijveld replied:
Looks like that works as well, I've had to add a line to polyfill File as well, and I had to disable eslint for those two lines to allow the conversion to any (the types of the nodejs ones and the browser ones technically don't match I believe). Changed in [c81b854](https://github.com/kiesraad/abacus/pull/1406/commits/c81b8546ff6acbdab351232b3fccda2db862c992)

Lionqueen94 commented on backend/src/election/api.rs:
```diff
@@ -0,0 +1,175 @@
+use super::structs::Election;
+#[cfg(feature = "dev-database")]
+use crate::audit_log::{AuditEvent, AuditService};
+use crate::{APIError, eml::EMLImportError};
+use crate::{
+    authentication::{Admin, User},
+    election::{ElectionRequest, repository::Elections},
+    eml::{EML110, EMLDocument, RedactedEmlHash},
+    polling_station::{PollingStation, repository::PollingStations},
+    {AppState, ErrorResponse},
+};
+#[cfg(feature = "dev-database")]
+use axum::http::StatusCode;
+use axum::{
+    Json,
+    extract::{Path, State},
+};
+use quick_xml::{DeError, SeError};
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+use utoipa_axum::{router::OpenApiRouter, routes};
```
```suggestion
#[cfg(feature = "dev-database")]
use axum::http::StatusCode;
use axum::{
    Json,
    extract::{Path, State},
};
use quick_xml::{DeError, SeError};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use utoipa_axum::{router::OpenApiRouter, routes};

use super::{ElectionRequest, repository::Elections, structs::Election};
#[cfg(feature = "dev-database")]
use crate::audit_log::{AuditEvent, AuditService};
use crate::{APIError, eml::EMLImportError};
use crate::{
    AppState, ErrorResponse
    authentication::{Admin, User},
    eml::{EML110, EMLDocument, RedactedEmlHash},
    polling_station::{PollingStation, repository::PollingStations},
};
```

praseodym commented:
> AFAIK the main reason for this check is so the user pays extra attention to the SHA hash, so they do a double check they upload the correct file. If there also security concerns for this check, then it's probably a good idea to put in a little more effort in the backend code so it always returns the same stubs. Let me know what you think and I'm also curious of @praseodym's thoughts on this matter.

It is, but at the same time it feels strange to reveal all parts of the hash if the same file is re-uploaded. Maybe the hash itself could be the seed to the random function, so that the redacted parts are the same for every file?

praseodym commented on frontend/src/features/election_management/components/CheckElectionDefinition.tsx:
```diff
@@ -0,0 +1,84 @@
+        <p>
+          <strong>{election.name}</strong>
+          <br />
+          <span className="capitalize">{formatDateFull(new Date(election.election_date))}</span>
```
In Dutch, names of weekdays and months should not be capitalized, so I think we should remove this `capitalize` class:
<img width="542" alt="Screenshot 2025-04-29 at 15 50 43" src="https://github.com/user-attachments/assets/e630014a-208f-400f-b7ae-3dfce9e840a3" />

cikzh replied:
It should capitalize only the first word of the sentence, I'll see if I can get it to only capitalize the first word.

praseodym commented:
I just noticed that the 'Verkiezing toevoegen' button appears for the coordinator role even though it is supposed to be only for the administrator, see our [authorization matrix](https://github.com/kiesraad/abacus/blob/main/documentatie/use-cases/autorisatiematrix.md).

cikzh commented:
> It is, but at the same time it feels strange to reveal all parts of the hash if the same file is re-uploaded. Maybe the hash itself could be the seed to the random function, so that the redacted parts are the same for every upload of the same file?

[f83445c](https://github.com/kiesraad/abacus/pull/1406/commits/f83445cf28c66cc65056b7742ddd0d95c4c821a8) adds an implementation for this

Lionqueen94 commented on backend/src/eml/tests/eml110a_test.eml.xml:
```diff
@@ -7,7 +7,7 @@
         <Election>
             <ElectionIdentifier Id="GR2022_Amsterdam">
-                <ElectionName>Gemeenteraad Amsterdam 2022</ElectionName>
+                <ElectionName>Gemeenteraad lkmsflkgmsfdklgjbfnsklfdjnAmsterdam 2022</ElectionName>
```
Was this an accidental change?

cikzh commented:
> I just noticed that the 'Verkiezing toevoegen' button appears for the coordinator role even though it is supposed to be only for the administrator, see our [authorization matrix](https://github.com/kiesraad/abacus/blob/main/documentatie/use-cases/autorisatiematrix.md).

Fixed it in [44f5560](https://github.com/kiesraad/abacus/pull/1406/commits/44f5560ef51c99bd2848c6c70cebebd38053662e).

cikzh commented on backend/src/eml/hash.rs:
```diff
@@ -0,0 +1,101 @@
+use rand::{SeedableRng, rngs::SmallRng};
```
No, it is used by `SmallRng::from_seed`. I've move the import to the function body though, which expresses the fact a little better hopefully: [fb267fa](https://github.com/kiesraad/abacus/pull/1406/commits/fb267fa5a6b47d8c800ba7faa521a14f9e6c7b1d)

Lionqueen94 commented on frontend/src/features/election_overview/components/OverviewPage.test.tsx:
```diff
@@ -47,4 +48,32 @@ describe("OverviewPage", () => {
+
+    // Wait for the page to be loaded
+    expect(await screen.findByRole("heading", { level: 1, name: "Beheer verkiezingen" })).toBeVisible();
+    expect(await screen.findByText("Verkiezing toevoegen")).toBeVisible();
```
```suggestion
    expect(await screen.findByRole("button", { name: "Verkiezing toevoegen" })).toBeInTheDocument();
```

cikzh replied:
Fixed in [d56f6f0](https://github.com/kiesraad/abacus/pull/1406/commits/d56f6f05afd85fd836bfb7ff01fbe11e59fd5f27)

Lionqueen94 commented on frontend/src/features/election_overview/components/OverviewPage.test.tsx:
```diff
@@ -47,4 +48,32 @@ describe("OverviewPage", () => {
+
+    // Wait for the page to be loaded
+    expect(await screen.findByRole("heading", { level: 1, name: "Beheer verkiezingen" })).toBeVisible();
+    expect(screen.queryByText("Verkiezing toevoegen")).not.toBeInTheDocument();
```
```suggestion
    expect(await screen.queryByRole("button", { name: "Verkiezing toevoegen" })).not.toBeInTheDocument();
```

cikzh replied:
Fixed in [d56f6f0](https://github.com/kiesraad/abacus/pull/1406/commits/d56f6f05afd85fd836bfb7ff01fbe11e59fd5f27)
